apiserver/facades/client/charms/register.go

```diff
@@ -19,47 +21,46 @@ import (
 // Register is called to expose a package of facades onto a given registry.
 func Register(registry facade.FacadeRegistry) {
 	registry.MustRegister("Charms", 7, func(stdCtx context.Context, ctx facade.ModelContext) (facade.Facade, error) {
-		return newFacadeV7(ctx)
+		return newFacadeV7(stdCtx, ctx)
 	}, reflect.TypeOf((*APIv7)(nil)))
 }
 
-func newFacadeV7(ctx facade.ModelContext) (*APIv7, error) {
-	api, err := newFacadeBase(ctx)
+func newFacadeV7(stdCtx context.Context, ctx facade.ModelContext) (*APIv7, error) {
+	api, err := makeFacadeBase(stdCtx, ctx)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	return &APIv7{api}, nil
 }
 
-// newFacadeBase provides the signature required for facade registration.
-func newFacadeBase(ctx facade.ModelContext) (*API, error) {
+// makeFacadeBase provides the signature required for facade registration.
+func makeFacadeBase(stdCtx context.Context, ctx facade.ModelContext) (*API, error) {
 	authorizer := ctx.Auth()
 	if !authorizer.AuthClient() {
 		return nil, apiservererrors.ErrPerm
 	}
 
 	st := ctx.State()
-	m, err := st.Model()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-
 	commonState := &charmscommon.StateShim{State: st}
 	charmInfoAPI, err := charmscommon.NewCharmInfoAPI(commonState, authorizer)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
+	modelInfo, err := ctx.ServiceFactory().ModelInfo().GetModelInfo(stdCtx)
+	if err != nil {
+		return nil, fmt.Errorf("getting model info: %w", err)
+	}
 	return &API{
 		charmInfoAPI:       charmInfoAPI,
 		authorizer:         authorizer,
 		backendState:       newStateShim(st),
-		backendModel:       m,
+		modelConfigService: ctx.ServiceFactory().Config(),
 		charmhubHTTPClient: ctx.HTTPClient(facade.CharmhubHTTPClient),
 		newRepoFactory: func(cfg services.CharmRepoFactoryConfig) corecharm.RepositoryFactory {
 			return services.NewCharmRepoFactory(cfg)
 		},
-		tag:             m.ModelTag(),
+		tag:             names.NewModelTag(modelInfo.UUID.String()),
```

Comment:
Small nit...

I dislike that we are using tags at this level. We should ideally just be passing the model id around and only going to tags on and off the wire. Up to you if you want to squash this here and now.

Reply:
don't really see the point tbh, the only usage is to check `authorizer.HasPermission` and for that it has to be a tag.